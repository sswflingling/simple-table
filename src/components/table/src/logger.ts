// 封装console
function createLog(): Console {
    // 创建一个对象 存储console里面的值，并且全都是非必填
    let logs:Partial<Console> = {};
    // 循环遍历console对象
    for (const k in console) {
        const key = k as keyof Console
        // 判断console中是否存在key
        if (Object.prototype.hasOwnProperty.call(console, key)) {
            
            const method:(...abc:any) => void = console[key ];

            // import.meta.env.DEV: 判断是否为开发环境
            logs[key] = (...abc) => import.meta.env.DEV && method(...abc);
        }
    }
    return logs as Console;
}

export const log = createLog()