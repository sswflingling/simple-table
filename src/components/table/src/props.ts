import type { ColumnsType } from "./types/index";
import type { PropType, ExtractPropTypes } from "vue";

export const basicProps = {
    data: {
        type: Array as PropType<{ [x: string]: any }[]>,
        default: ()=> []
    },
    columns: {
        type: Array as PropType<ColumnsType[]>,
        default: ()=> [],
        required: true
    }
};

export type BasicProps = Readonly<ExtractPropTypes<typeof basicProps>>;
