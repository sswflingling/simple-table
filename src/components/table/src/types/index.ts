import type { VNode } from "vue"

export type Directions = 'ascend' | 'descend' // ascend:顺序 descend:倒叙

// 表头数据类型
export interface ColumnsType {
    title: string;
    key?: string;
    width?: number;
    align?: 'left' | 'right' | 'center';
    isColResize?: boolean; // 是否可拉动列宽
    children?: ColumnsType[];
    render?: (obj: { content: string; index: number; row: any; col: ColumnsTheadType; }) => string | number | VNode; // 
    sorter?: ((directions: Directions, a: any, b: any) => number) | boolean;
    fixed?: 'left' | 'right';
}

export interface ColumnsTheadType extends ColumnsType {
    level: number;
    classResize: string;
    parent?: ColumnsTheadType;
    children?: ColumnsTheadType[];
}