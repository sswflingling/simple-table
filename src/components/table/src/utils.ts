let styleIndex = 0

// 获取类当前下标
function getCssRuleIndex(className: string): number {
    const cssRules = document.styleSheets[0].cssRules
    for (const key in cssRules) {
        if ((cssRules.item(key as unknown as number) as any)?.selectorText == '.' + className) {
            return key as unknown as number
        }
    }
    return -1
}

// 动态插入css样式
export function insertRule(className: string, style: string) {
    if (getCssRuleIndex(className) !== -1) return;
    styleIndex += 1
    document.styleSheets[0].insertRule('.' + className + ' ' + style, styleIndex)
}
// 动态更改css样式
export function updateRule(className: string, obj: { [x: string]: string }) {
    const index = getCssRuleIndex(className)
    if (index === -1) return;
    const domStyle = (document.styleSheets[0].cssRules[index] as any).style
    for (const key in obj) {
        domStyle[key] = obj[key]
    }
}
