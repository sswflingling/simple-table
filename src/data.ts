import type { ColumnsType } from "@/components/table";

import { h } from "vue"

export const columns: ColumnsType[] = [
    {
        title: "ID",
        key: "key",
        width: 100
    },
    {
        title: "用户名",
        key: "name",
        width: 100,
        fixed: "left",
        align: "center",
        isColResize: true,
        sorter: true
    },
    {
        title: "Other",
        children: [
            {
                title: "Age",
                key: "age",
                width: 200,
                render: ({ content }) => h('div', {}, content),
                sorter: true
            },
            {
                title: "Address",
                children: [
                    {
                        title: "Street",
                        key: "street",
                        width: 200
                    },
                    {
                        title: "Block",
                        children: [
                            {
                                title: "BuildingBuildingBuildingBuildingBuildingBuilding",
                                key: "building",
                                width: 100,
                                isColResize: true
                            },
                            {
                                title: "Door No.",
                                key: "number",
                                width: 100,
                                sorter: (order, a, b) => {
                                    if (order === "ascend") {
                                        return a.number - b.number;
                                    } else {
                                        return b.number - a.number;
                                    }
                                }
                            }
                        ]
                    }
                ]
            }
        ]
    },
    {
        title: "Company",
        children: [
            {
                title: "Company Address",
                key: "companyAddress",
                width: 200,
                isColResize: true,
            },
            {
                title: "Company Name",
                key: "companyName",
            },
        ]
    },
    {
        title: "userName",
        key: "user.name",
        width: 100,
        fixed: "right",
        isColResize: true,
    },
    {
        title: "Gender",
        key: "gender",
        width: 80,
        fixed: "right",
        isColResize: true,
    },
];

export const data = [...Array(100)].map((item, i) => ({
    key: i,
    name: Math.random().toString(30),
    age: Math.floor(Math.random() * 100),
    street: "Lake Park",
    building: "C",
    number: Math.floor(Math.random() * 200),
    companyAddress:
        "Lake Street 42 BuildingBuildingBuildingBuildingBuildingBuilding",
    companyName: "SoftLake Co",
    gender: "M",
    user: {
        name: "小蛮"
    }
}));