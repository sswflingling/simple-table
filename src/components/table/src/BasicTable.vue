<template>
    <div class="basic-table">
        <div class="table">
            <table class="table-box table-border"
                   ellspacing="0"
                   cellpadding="0"
                   border="0">
                <Thead />
                <Tbody />
            </table>
        </div>

    <div class="pagination">4555</div>
    </div>
</template>


<script lang="ts" setup>

import Thead from "./components/Thead.vue"; //表头组件
import Tbody from "./components/Tbody.vue"; // 表格内容组件

import { basicProps } from "./props";
import { useBasicTable } from "./hooks/useBasicTable";

const props = defineProps(basicProps);

useBasicTable(props);

</script>
<style lang="less" scoped>
.basic-table {
    width: 100%;
    height: 100%;
    overflow: hidden;
    display: flex;
    flex-direction: column;
    .table {
        height: 100%;
        flex: 1;
        overflow: hidden;
        overflow: auto;
        .table-box {

            &:deep(.table-thead) {
                flex-shrink: 0;
            }

            &:deep(.table-tbody) {
                flex: 1;
                overflow: hidden;
                overflow-y: auto;
                position: relative;
            }
        }
    }

    .pagination {
        flex-shrink: 0;
    }
    .table-border {
        border: 1px solid #eee;
        border-bottom: none;
        border-right: none;

        &:deep(th),
        &:deep(td) {
            border: 1px solid #eee;
            border-top: none;
            border-left: none;
        }
    }
}
</style>
