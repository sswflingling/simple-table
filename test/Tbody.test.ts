import { MountingOptions, mount, shallowMount } from "@vue/test-utils";
import { expect, test, describe } from "vitest";

import type { BasicProps } from "../src/components/table/src/props";

import Table from '../src/App.vue'
import BasicTable from '../src/components/table/src/BasicTable.vue'
import { columns, data } from "../src/data";

// let _wrapper;
// beforeEach(async () => {
//     // 在每次测试运行之前清除模拟并添加一些测试数据
// })

describe('Table', () => {
    describe('basic work', () => {
        const TableMount = (options?: MountingOptions<Partial<BasicProps>>) =>  {
            const { props } = options || {};
            const mergedOptions = {
                props: { columns, data, ...props }
            } as MountingOptions<BasicProps>
            return mount(Table, mergedOptions)
        }
        

        test('tbody work', async () => {
            const wrapper = TableMount();
            expect(wrapper.find('tbody').classes()).toContain('table-tbody');
            expect(wrapper.find('tbody').findAll('tr')[0].classes()).toContain('table-tbody-tr');
        })

        test('thead work', () => {
            const wrapper = TableMount();
            expect(wrapper.find('thead').classes()).toContain('table-thead');
        })

        // test('tableCell work', () => {
        //     const wrapper = TableMount();
        //     expect(wrapper.find('thead').classes()).toContain('table-thead');
        // })
        
        // test('sort work', () => {
            // const wrapper = TableMount();
            // expect(wrapper.find('th').classes()).toContain('cursor-pointer');
        // })

        test('pagination work', async() => {
            const wrapper = TableMount();
            const paginationWrapper = shallowMount(BasicTable,{
                props: {
                    pageSize: 10,
                    total: 50
                }
            })
            expect(paginationWrapper.html()).toMatchSnapshot();
            // expect(wrapper.find('tbody').findAll('tr').length).toBe(10)

            // await paginationWrapper.setProps({pageSize: 20 })

            // expect(wrapper.find('tbody').findAll('tr').length).toBe(20)
            // expect(wrapper.findAll('.ix-pagination-item').length).toEqual(7)
        })

    })

    
})
