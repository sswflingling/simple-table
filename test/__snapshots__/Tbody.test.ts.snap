// Vitest Snapshot v1

exports[`Table > basic work > pagination work 1`] = `
"<div class=\\"basic-table\\" pagesize=\\"10\\" total=\\"50\\" data-v-170bffdc=\\"\\">
  <div class=\\"table\\" data-v-170bffdc=\\"\\">
    <table class=\\"table-box table-border\\" cellspacing=\\"0\\" cellpadding=\\"0\\" border=\\"0\\" data-v-170bffdc=\\"\\">
      <thead-stub data-v-170bffdc=\\"\\"></thead-stub>
      <tbody-stub data-v-170bffdc=\\"\\"></tbody-stub>
    </table>
  </div>
  <div class=\\"pagination\\" data-v-170bffdc=\\"\\">
    <pagination-stub currentpage=\\"1\\" pagesize=\\"10\\" pagesizes=\\"10,20,30,40\\" total=\\"0\\" data-v-170bffdc=\\"\\"></pagination-stub>
  </div>
</div>"
`;
